src/devices/cddasongloader.cpp

```diff
@@ -55,17 +68,88 @@ bool CddaSongLoader::IsActive() const { return loading_future_.isRunning(); }
 void CddaSongLoader::LoadSongs() {
   // only dispatch a new thread for loading tracks if not already running.
   if (!IsActive()) {
+    disc_ = Disc();
     loading_future_ =
         QtConcurrent::run(this, &CddaSongLoader::LoadSongsFromCdda);
   }
 }
 
+bool CddaSongLoader::ParseSongTags(SongList& songs, GstTagList* tags) {
+  //// cdiocddasrc reads cd-text with following mapping from cdio
+  ///
+  /// DISC LEVEL :
+  /// CDTEXT_FIELD_PERFORMER -> GST_TAG_ALBUM_ARTIST
+  /// CDTEXT_FIELD_TITLE -> GST_TAG_ALBUM
+  /// CDTEXT_FIELD_GENRE -> GST_TAG_GENRE
+  ///
+  /// TRACK LEVEL :
+  /// CDTEXT_FIELD_PERFORMER -> GST_TAG_ARTIST
+  /// CDTEXT_FIELD_TITLE -> GST_TAG_TITLE
+
+  guint track_number;
+  if (!gst_tag_list_get_uint(tags, GST_TAG_TRACK_NUMBER, &track_number)) {
+    qLog(Error) << "Track tags do not contain track number!";
+    return false;
+  }
+
+  Q_ASSERT(track_number != 0u);
+  Q_ASSERT(static_cast<int>(track_number) <= songs.size());
+  Song& song = songs[static_cast<int>(track_number - 1)];
+
+  qLog(Debug) << gst_tag_list_to_string(tags);
+
+  bool has_loaded_tags = false;
+
+  gchar* buffer = nullptr;
+  if (gst_tag_list_get_string(tags, GST_TAG_ALBUM, &buffer)) {
+    has_loaded_tags = true;
+    song.set_album(QString::fromUtf8(buffer));
+    g_free(buffer);
+  }
+
+  if (gst_tag_list_get_string(tags, GST_TAG_ALBUM_ARTIST, &buffer)) {
+    has_loaded_tags = true;
+    song.set_albumartist(QString::fromUtf8(buffer));
+    g_free(buffer);
+  }
+
+  if (gst_tag_list_get_string(tags, GST_TAG_GENRE, &buffer)) {
+    has_loaded_tags = true;
+    song.set_genre(QString::fromUtf8(buffer));
+    g_free(buffer);
+  }
+
+  if (gst_tag_list_get_string(tags, GST_TAG_ARTIST, &buffer)) {
+    has_loaded_tags = true;
+    song.set_artist(QString::fromUtf8(buffer));
+    g_free(buffer);
+  }
+
+  if (gst_tag_list_get_string(tags, GST_TAG_TITLE, &buffer)) {
+    has_loaded_tags = true;
+    song.set_title(QString::fromUtf8(buffer));
+    g_free(buffer);
+  }
+
+  guint64 duration;
+  if (gst_tag_list_get_uint64(tags, GST_TAG_DURATION, &duration)) {
+    has_loaded_tags = true;
+    song.set_length_nanosec(duration);
+  }
+
+  song.set_track(track_number);
+  song.set_id(track_number);
+  song.set_filetype(Song::Type_Cdda);
+  song.set_valid(true);
+  song.set_url(GetUrlFromTrack(track_number));
+  return has_loaded_tags;
+}
+
 void CddaSongLoader::LoadSongsFromCdda() {
   if (!may_load_) return;
```

Comment:
Is the object leaked in this case?

Reply:
Which object exactly do you mean? I had a brief look at SongLoader::LoadLocalPartial but can't really see how that interacts with this as it seems concerned with local files, not CDs. Correct me if I'm wrong please.

Reply:
Sorry, I read the wrong method name. I meant SongLoader::LoadAudioCD(). That creates an instance of CddaSongLoader and relies on a signal to tell it when it can delete the object. We should make sure your finished signal is emitted in all cases.

Reply:
Ah, I see. I have added a `Finished` signal to `CddaSongLoader` according to your suggestion below and double checked that every branch that ends the loading process in `CddaSongLoader`, abort or success, emits `Finished`, so I think that the changes work well with `SongLoader::LoadAudioCD`. It is a bit hard to test with 100% certainty, though.